### PR TITLE
RuboCop changes.

### PR DESCRIPTION
**Pull request by iamAzeem**
RuboCop changes applied.
Along with RuboCop's [defaults](https://github.com/rubocop-hq/rubocop/blob/master/config/default.yml), some exceptions have been added to [.rubocop.yml](.rubocop.yml).

For now, new cops are not entertained.

Signed-off-by: Azeem Sajid <<azeem.sajid@gmail.com>>

### CODE REVIEW

File: lib/kaitai/console_windows.rb [hunk 1]
```diff
@@ -1,266 +1,262 @@
...
+      goto(x, y)
+      puts "[ #{caption} ]"
+    end
 
-  # Regexp borrowed from
-  # http://stackoverflow.com/questions/170956/how-can-i-find-which-operating-system-my-ruby-program-is-running-on
-  @@is_windows = (RUBY_PLATFORM =~ /cygwin|mswin|mingw|bccwin|wince|emx/) ? true : false
+    # Regexp borrowed from
+    # http://stackoverflow.com/questions/170956/how-can-i-find-which-operating-system-my-ruby-program-is-running-on
+    @@is_windows = RUBY_PLATFORM =~ /cygwin|mswin|mingw|bccwin|wince|emx/ ? true : false
```
Comment (ams-tschoening) [inline lib/kaitai/console_windows.rb:245]: Brackets should be kept in my  opinion, is more readable with them.

File: lib/kaitai/tui.rb [hunk 2]
```diff
@@ -1,113 +1,100 @@
...
+      goto(x, y)
+      puts "[ #{caption} ]"
+    end
 
-  # Regexp borrowed from
-  # http://stackoverflow.com/questions/170956/how-can-i-find-which-operating-system-my-ruby-program-is-running-on
-  @@is_windows = (RUBY_PLATFORM =~ /cygwin|mswin|mingw|bccwin|wince|emx/) ? true : false
+    # Regexp borrowed from
+    # http://stackoverflow.com/questions/170956/how-can-i-find-which-operating-system-my-ruby-program-is-running-on
+    @@is_windows = RUBY_PLATFORM =~ /cygwin|mswin|mingw|bccwin|wince|emx/ ? true : false
```
Comment (ams-tschoening) [inline lib/kaitai/tui.rb:93]: Brackets should be kept in my  opinion, is more readable with them.

File: kaitai-struct-visualizer.gemspec [hunk 3]
```diff
@@ -1,9 +1,9 @@
-# -*- encoding: utf-8 -*-
+# frozen_string_literal: true
 
-require File.expand_path("../lib/kaitai/struct/visualizer/version", __FILE__)
+require File.expand_path('lib/kaitai/struct/visualizer/version', __dir__)
```
Comment (ams-tschoening) [inline kaitai-struct-visualizer.gemspec:3]: Just want to document that `__dir__` doesn't seem to depend on the current working dir and should therefore be safe and what's actually done in the code before:

    File.dirname(File.realpath(__FILE__))

https://stackoverflow.com/a/15190725/2055163
https://bugs.ruby-lang.org/issues/7975
https://web.archive.org/web/20100414044710/http://www.ruby-forum.com/topic/193674

**Comment by ams-tschoening** [comment]
Is that a one time formatting? If RoboCup supports some kind of config files with what it has done, that should be committed with the repo, so that one easily knows and can change those configs. Pretty much like what is done with `.editorconfig` or [other style formatter configs](https://github.com/apache/logging-log4cxx/blob/master/.astylerc).

File: lib/kaitai/console_windows.rb (hunk 1 above)
Comment (iamAzeem) [reply lib/kaitai/console_windows.rb]: Incorporated this change with https://github.com/kaitai-io/kaitai_struct_visualizer/pull/37/commits/049dadb6fc63330387c05c68aac4d3917896dd17.

File: lib/kaitai/tui.rb (hunk 2 above)
Comment (iamAzeem) [reply lib/kaitai/tui.rb]: Incorporated with https://github.com/kaitai-io/kaitai_struct_visualizer/pull/37/commits/049dadb6fc63330387c05c68aac4d3917896dd17.

**Comment by iamAzeem** [comment]
@ams-tschoening: Yes, this is a one time formatting. Depending on the new rules, the formatting may change with time.

I've added [.rubocop.yml](.rubocop.yml) with https://github.com/kaitai-io/kaitai_struct_visualizer/pull/37/commits/049dadb6fc63330387c05c68aac4d3917896dd17.

The [default configuration](https://github.com/rubocop-hq/rubocop/blob/master/config/default.yml) is hosted by RuboCop itself. The [.rubocop.yml](.rubocop.yml) contains the exclusions for certain rules that would require some code refactoring first to apply on the codebase. Those rules can be changed later if required.

**Comment by ams-tschoening** [comment]
I've pulled the PR to test with my available Windows-setup and things fail:

	C:/Users/tschoening/Documents/Eclipse/Ruby GitHub/kaitai_struct_visualizer/lib/kaitai/tui.rb:14: warning: Win32API is deprecated after Ruby 1.9.1; use fiddle directly instead
	Traceback (most recent call last):
	        11: from C:/Users/tschoening/Documents/Eclipse/Java Sm-Mtg bug_2590/../Ruby GitHub/kaitai_struct_visualizer/bin/ksv:58:in `<main>'
	        10: from C:/Users/tschoening/Documents/Eclipse/Ruby GitHub/kaitai_struct_visualizer/lib/kaitai/struct/visualizer/visualizer.rb:20:in `run'
	         9: from C:/Users/tschoening/Documents/Eclipse/Ruby GitHub/kaitai_struct_visualizer/lib/kaitai/struct/visualizer/tree.rb:162:in `redraw'
	         8: from C:/Ruby/2.6.5/lib/ruby/2.6.0/benchmark.rb:308:in `realtime'
	         7: from C:/Users/tschoening/Documents/Eclipse/Ruby GitHub/kaitai_struct_visualizer/lib/kaitai/struct/visualizer/tree.rb:163:in `block in redraw'
	         6: from C:/Ruby/2.6.5/lib/ruby/2.6.0/forwardable.rb:230:in `clear'
	         5: from C:/Users/tschoening/Documents/Eclipse/Ruby GitHub/kaitai_struct_visualizer/lib/kaitai/console_windows.rb:69:in `clear'
	         4: from C:/Ruby/2.6.5/lib/ruby/2.6.0/Win32API.rb:38:in `call'
	         3: from C:/Ruby/2.6.5/lib/ruby/2.6.0/Win32API.rb:38:in `each_with_index'
	         2: from C:/Ruby/2.6.5/lib/ruby/2.6.0/Win32API.rb:38:in `each'
	         1: from C:/Ruby/2.6.5/lib/ruby/2.6.0/Win32API.rb:39:in `block in call'
	C:/Ruby/2.6.5/lib/ruby/2.6.0/Win32API.rb:39:in `pack': can't modify frozen String (FrozenError)

**Comment by iamAzeem** [comment]
@ams-tschoening: Looking into it. Can you please test on Linux also and observe the behavior?

**Comment by ams-tschoening** [comment]
Not really, I don't have a working Linux-setup for that project at all. Doesn't everyone else uses Linux and MAC? :-)

**Comment by iamAzeem** [comment]
Well, almost everyone. :-)
Please remove the first line `# frozen_string_literal: true` from `tui.rb` and test again.

**Comment by ams-tschoening** [comment]
That file was not enough, the line needed to be removed in `node.rb` as well. Additionally, automatically applied changes in `tree.rb` failed:

      if @cur_line - @cur_shift < 0
        @cur_shift = @cur_line

vs.

    @cur_shift = @cur_line if @cur_line - @cur_shift.negative?

The new line lead to type errors from boolean to integer, the `if` was used as `if @cur_line - (@cur_shift < 0)` instead of `if (@cur_line - @cur_shift) < 0`.

tree rb
[fixes.patch.txt](https://github.com/kaitai-io/kaitai_struct_visualizer/files/4877563/fixes.patch.txt)

**Comment by iamAzeem** [comment]
`console_windows.rb`:
Add the first line `# frozen_string_literal: true`.

Change line # 68 from:
```
num_written = 'XXXX'
```
to
```
num_written = 'XXXX'.dup
```

And, test again.

**Comment by ams-tschoening** [comment]
Wasn't enough, `node.rb` needed to be changed as well:

    r = ''.dup

Does it really make sense to explicitly freeze strings just to copy them afterwards?

[fixes.patch.txt](https://github.com/kaitai-io/kaitai_struct_visualizer/files/4877757/fixes.patch.txt)

**Comment by iamAzeem** [comment]
@ams-tschoening: Only a few of these API calls and string usages need mutable strings. From the performance point-of-view, it does make sense to employ [copy-on-write](https://en.wikipedia.org/wiki/Copy-on-write) idiom.

**Review by generalmimon — CHANGES_REQUESTED** [review]

**Review by generalmimon — APPROVED** [review]
LGTM, thanks for the contribution!

**Comment by abitrolly** [comment]
I believe `ksv` is unable to process keypresses after this merge. When I build Docker image without this commit, it works without problems.

**Comment by generalmimon** [comment]
> I believe `ksv` is unable to process keypresses after this merge. When I build Docker image without this commit, it works without problems.

Which OS, Ruby version and console do you use?

**Comment by abitrolly** [comment]
@generalmimon this one
```
podman run -v "$(pwd):/share":Z -it yakshaveinc/ksv:rubocop-37 v3.snap squashfs_superblock.ksy
```
No keys are working. Although the bottom line updates on each keypress.

**Comment by abitrolly** [comment]
```
# cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
# ruby --version
ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-linux]
```

**Comment by generalmimon** [comment]
> ```bash
> podman run -v "$(pwd):/share":Z -it yakshaveinc/ksv:rubocop-37 v3.snap squashfs_superblock.ksy
> ```

@abitrolly Can I ask you please to try it outside the Docker container?

**Comment by abitrolly** [comment]
@generalmimon previous commit inside the same docker container works fine.
```
podman run -v "$(pwd):/share":Z -it yakshaveinc/ksv v3.snap squashfs_superblock.ksy
```
Installing Ruby with dependencies was a mess last time I tried. That's why I prefer Docker.

**Pull request merged**